substrate/tendermint/machine/src/lib.rs

```diff
@@ -0,0 +1,620 @@
+            match step {
+              Step::Propose => {
+                // Slash the validator for not proposing when they should've
+                self.slash(
+                  self.weights.proposer(self.block.number, self.block.round().number)
+                ).await;
+                self.broadcast(Data::Prevote(None));
+              },
+              Step::Prevote => self.broadcast(Data::Precommit(None)),
+              Step::Precommit => {
+                self.round(RoundNumber(self.block.round().number.0 + 1), None);
+                continue;
+              }
+            }
+          }
+          None
+        },
+
+        // Handle any received messages
+        msg = self.msg_recv.next() => {
+          if let Some(msg) = msg {
+            if !msg.verify_signature(&self.validators) {
+              continue;
+            }
+            Some((false, msg.msg))
+          } else {
+            break;
+          }
+        }
+      } {
+        let res = self.message(msg.clone()).await;
+        if res.is_err() && broadcast {
+          panic!("honest node had invalid behavior");
+        }
+
+        match res {
+          Ok(None) => (),
+          Ok(Some(block)) => {
+            let mut validators = vec![];
+            let mut sigs = vec![];
+            // Get all precommits for this round
+            for (validator, msgs) in &self.block.log.log[&msg.round] {
+              if let Some(Data::Precommit(Some((id, sig)))) = msgs.get(&Step::Precommit) {
+                // If this precommit was for this block, include it
+                if id == &block.id() {
+                  validators.push(*validator);
+                  sigs.push(sig.clone());
+                }
+              }
+            }
+
+            let commit = Commit {
+              end_time: self.block.end_time[&msg.round].canonical(),
+              validators,
+              signature: N::SignatureScheme::aggregate(&sigs),
+            };
+            debug_assert!(self.network.verify_commit(block.id(), &commit));
+
+            let proposal = self.network.add_block(block, commit).await;
+            self.reset(msg.round, proposal).await;
+          }
+          Err(TendermintError::Malicious(validator)) => {
+            self.slash(validator).await;
+          }
+          Err(TendermintError::Temporal) => (),
+        }
+
+        if broadcast {
+          let sig = self.signer.sign(&msg.encode()).await;
+          self.network.broadcast(SignedMessage { msg, sig }).await;
+        }
+      }
+    }
+  }
+
+  // Returns Ok(true) if this was a Precommit which had its signature validated
+  // Returns Ok(false) if it wasn't a Precommit or the signature wasn't validated yet
+  // Returns Err if the signature was invalid
+  fn verify_precommit_signature(
+    &self,
+    sender: N::ValidatorId,
+    round: RoundNumber,
+    data: &DataFor<N>,
+  ) -> Result<bool, TendermintError<N::ValidatorId>> {
+    Ok(if let Data::Precommit(Some((id, sig))) = data {
```

Comment:
This just returns Ok so long as nothing in the function errors. The alternative is

```
let res = if { ...
Ok(res)
````

I'm not a fan of the pointless assignment, especially since this isn't causing line breaks, yet I'd do it to squash this.